src/main/java/com/commercetools/pspadapter/notification/processor/impl/PaymentSaleNotificationProcessorBase.java

```diff
@@ -35,19 +46,70 @@ public boolean canProcess(@Nonnull Event event) {
                 .equalsIgnoreCase(event.getEventType());
     }
 
-    protected AddTransaction createAddTransactionAction(@Nonnull Event event,
-                                                        @Nonnull TransactionType transactionType) {
+    protected List<UpdateAction<Payment>> createUpdatePaymentActions(@Nonnull Payment ctpPayment, @Nonnull Event event) {
+        String resourceId = getResourceId(event);
+        return findTransactionByInteractionId(ctpPayment.getTransactions(), resourceId)
+                .map(txn -> processNotificationForTransaction(ctpPayment, event, resourceId, txn))
+                .orElseGet(() -> createAddTransactionActionList(event, getExpectedTransactionType()));
+    }
+
+    private List<UpdateAction<Payment>> processNotificationForTransaction(@Nonnull Payment ctpPayment, @Nonnull Event event,
+                                                                          @Nonnull String resourceId, @Nonnull Transaction txn) {
+        if (!txn.getType().equals(getExpectedTransactionType())) {
+            logger.info("Found txn paymentId=[{}] with corresponding resourceId={},"
+                            + " but transactionType=[{}] is not expectedTransactionType=[{}]."
+                            + " Will create new transaction for the eventId=[{}]",
+                    ctpPayment.getId(), resourceId, getExpectedTransactionType(), event.getId());
+            return createAddTransactionActionList(event, getExpectedTransactionType());
+        } else if (isTxnAlreadyUpdated(txn)) {
+            // can't do Collections.emptyList() here because map() and generics
+            // together generates unexpected return results
+            return Collections.emptyList();
+        } else {
+            return createChangeTransactionStateActionList(txn);
+        }
+    }
+
+    @Nonnull
+    abstract protected TransactionType getExpectedTransactionType();
+
+    @Nonnull
+    abstract protected TransactionState getExpectedTransactionState();
+
+    protected List<UpdateAction<Payment>> createChangeTransactionStateActionList(Transaction txn) {
+        return Collections.singletonList(ChangeTransactionState.of(getExpectedTransactionState(), txn.getId()));
+    }
+
+    protected List<UpdateAction<Payment>> createAddTransactionActionList(@Nonnull Event event,
+                                                                         @Nonnull TransactionType transactionType) {
+        try {
+            Map resource = (Map) event.getResource();
+            String resourceId = (String) resource.get(ID);
+            Map amount = (Map) resource.get(AMOUNT);
+            BigDecimal total = new BigDecimal((String) amount.get(TOTAL));
```

Comment:
@lojzatran 
do you know that Paypal plus uses decimal major currency format with fixed 2 digits, when in CTP we should store in cents format without decimal values?

Reply:
#80